### ISSUE TITLE
Performance issue: MSAM Alarms table take very long to update (many alarms to be monitored)

### ISSUE THREAD

**Issue by chrisgel15** [aws-solutions/aws-media-services-application-mapper#65]
Hello, 

We have been developing a monitoring dashboard with MSAM for the past weeks and now that we make it to production environment we are facing difficulties with its performance.

### Environment
Our environment/deployments consists in the following: 

- 90 channels in Ireland and 90 channels in Frankurt --> 180 tiles in MSAM.
- Each channel has 2 MediaConnect instances, 1 MediaLive instance and 8 MediaPackage Endpoints instances.
- Each channel has almost 30 CloudWatch alarms associated with different metrics.

That gives us a total of 180 tiles * 30 = 5400 alarms that we are expecting to monitor within one instance of MSAM application.

Also:
- We are subscribing the resources to alarms and associating them with each tile through a bash script that provisions the DynamoDB tables (Alarms, Channel and Settings).
- We increased the Read/Write capactity to 40 units in Alarms, Channels and Events table as recommended in the documentation.

### Issue
We have found difficulties for MSAM to refresh the tiles in a timely fashion when mentioned environment was setup. We also performed several tests with different amount of alarms/tiles to understand this behavior and discover our bottleneck but so far we could not make it work accordingly.

The behavior that we see is that a group of alarms take longer to update/refresh its status. This time seems to increase with more alarms subscribed to tiles, so with our amount of alarms in the system this delay takes forever. This behavior is observed by looking at the Alarms table, column named "Updated", which does not change its value for this long time.

Below you can find a quick summary in graphs of the several tests that we performed.

_In the bubble graph below you can observe the little dots in the column of 60 minutes, which are pending alarms that were not updated for the past 40 minutes._

![image](https://user-images.githubusercontent.com/11690066/65542640-eb6a4c80-dee5-11e9-912c-8eb6366bbb47.png)

_Same test but with more tiles (40). Those upper bubbles are bigger because they represent more "not updated alarms"_
![image](https://user-images.githubusercontent.com/11690066/65542899-76e3dd80-dee6-11e9-8309-24580dd3f85a.png)

_A final test with only 60 tiles shows a group of alarms that was not updated for the past 120 minutes. That note that in our production scenario we have 180 tiles._
![image](https://user-images.githubusercontent.com/11690066/65543009-a72b7c00-dee6-11e9-84d4-463aa842c1de.png)

### Questions

1. Is MSAM tool designed to meet our requirements of monitoring ~5500 alarms and show updates in a timely fashion (5 minutes or less at a maximum)?. _If not, please tell us the maximum of alarms that we could monitor and its response time so we can make a decission._
2. Is there any setting that we could configure to drastically improve the performance?
3. Could you see any issues with our DB script provisioning that could affect the performance so much?
4. Could you suggest improvements to our design in order to find bottlenecks? _(we are analyzing dividing the instances and monitor only one region per instance, so it could have half of the tiles configured (~90))_

On a final note, any comments to help us improve the performance would be appreciated and let us know if you need more information regarding this issue.

**Comment by JimTharioAmazon** [comment]
Hi Christian, thank you for the detailed analysis of the problem. I just returned from holiday and need another day to review this and do my own investigation. If needed, we can arrange an online meeting to work on this collaboratively. I will respond again here in about 24 hours with my thoughts on how we might refactor the existing design and optimize for a large number of alarms.

**Comment by chrisgel15** [comment]
Hello Jim, thanks for the quick response.

Sure, take your time to review it and let me know how you can help us. Absolutely, we can arrange a meeting if you want.

Thanks!

**Comment by JimTharioAmazon** [comment]
Hi, I'm updating the thread here for the community after our external conversations yesterday.

Let's make sure your DynamoDB tables are using On Demand capacity. You can set that on existing tables under the Capacity tab or run the release template for the table and they will be updated.

I am found a couple of places in the Lambda code where I can optimize the DynamoDB interactions.

Because of the large number of alarms needing update, we may need to adapt the alarm update code to more of a continuous process, versus trying to squeeze it all in to a discrete time window before another periodic process is launched. I have a feeling we have multiple invocations of the same periodic process that's falling behind and essentially amplifying the latency each time. I'm continuing to look at this part, and will have another update by early next week.

**Comment by chrisgel15** [comment]
Hello Jim, just following up on this. Any updates on the modifications? Thanks!

**Comment by JimTharioAmazon** [comment]
Hi, there are a couple things happening.

1. We did not yet support the new JSON format for anomaly detection alarms. I was not aware these had a different format than the classic alarms. That is being fixed and I will have a new release out soon for that. Testing now.

2. We will need to deactivate the polling mechanism for alarm updates. We are capped at 9 TPS per account for DescribeAlarm calls. With that limit, it takes 10+ minutes to update every alarm in your configuration. This will slowly fall further behind since other resources in the account are likely checking on CloudWatch too. https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html

3. We will need to use a notification mechanism to receive alarm state changes. This means we'll need a SNS topic dedicated to this, have each alarm send OK and ALARM state changes to this topic, and that topic triggers a Lambda currently included in MSAM releases. This is a faster and more efficient means to keep MSAM's view current with the alarm state changes.

Once I have the changes to #1 complete, I will send out details for #2 & #3 in a document for you, and we can also talk through the changes needed on your side on a web meeting. If that looks good, I will include that here in the repo for the rest of the community.

**Comment by chrisgel15** [comment]
Hello Jim, thanks for the update! The design sound pretty well. 

I saw you released #1 changes yesterday, will try to apply then ASAP.

Do you have any ETA for changes #2 and #3? We would like to set expectations with our customer.

Thanks!

**Comment by JimTharioAmazon** [comment]
Hi, 1.4.2 will handle the new AD alarm JSON format. I will post a document today describing how to take advantage of high-performance alarm notifications. I need a few more hours to put that together.

**Comment by JimTharioAmazon** [comment]
Please review this new 1st draft document: https://github.com/awslabs/aws-media-services-application-mapper/blob/master/SCALING_UP_ALARMS.md

We can discuss here or on a web meeting Friday for questions/feedback if needed.

**Comment by chrisgel15** [comment]
Hello Jim, thank for the documentation. It looks pretty good. We have a couple of questions after reading it.

1. Do we need to wait for a code update on your side to start working on that configuration? 
2. We already have an SNS topic configured for our alarms, which points to a notification service via HTTP protocol. Can we add a subscription on it to point to the Lambda and reuse the same topic?

Thanks!
Christian

**Comment by JimTharioAmazon** [comment]
Hi,

1. I recommend you update if you are using alarms with anomaly detection. Previous versions of MSAM don't recognize the JSON response format of these alarms, and that can cause you problems.

2. Yes, you can create a trigger from the existing SNS topic to MSAM's Lambda. The CloudWatch alarms will need to notify on OK and Alarm states. MSAM will ignore Insufficient Data notifications in case you use that state for other purposes.

**Comment by JimTharioAmazon** [comment]
Hi Christian, I am checking in to see how this issue is progressing on your side. Let me know if we should keep this one open or close it. It is no problem to keep it open if you anticipate more work related to this issue. Thanks!

**Comment by chrisgel15** [comment]
Hello Jim, finally we completed our latest version deployment to production. We have some issues on previous deployment so completing this task was postponed a little bit.
We were able to update all of our alarms we are monitoring the tool. We will give you some updates when we have more information about its performance.

Once thing to point as feedback is that we have some alarms with a differente configuration on the `TreatMissingData` property, so they change from `Alarm` to `InsufficientData` and not to `OK` status. In the guide you share, you suggested to add notifications to `Alarm` and `OK` actions we have this edge case here. Whenever our alarm changes to `InsufficientData`, MSAM was not notified. Does it makes sense to you? Thanks
